# Dataset Load

In [1]:
from sklearn.datasets import load_wine
dataset = load_wine()

In [2]:
print(dataset.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

:Number of Instances: 178
:Number of Attributes: 13 numeric, predictive attributes and the class
:Attribute Information:
    - Alcohol
    - Malic acid
    - Ash
    - Alcalinity of ash
    - Magnesium
    - Total phenols
    - Flavanoids
    - Nonflavanoid phenols
    - Proanthocyanins
    - Color intensity
    - Hue
    - OD280/OD315 of diluted wines
    - Proline
    - class:
        - class_0
        - class_1
        - class_2

:Summary Statistics:

============================= ==== ===== ======= =====
                                Min   Max   Mean     SD
============================= ==== ===== ======= =====
Alcohol:                      11.0  14.8    13.0   0.8
Malic Acid:                   0.74  5.80    2.34  1.12
Ash:                          1.36  3.23    2.36  0.27
Alcalinity of Ash:            10.6  30.0    19.5   3.3
Magnesium:                    70.0 162.0    99.7  14.3

In [3]:
x, y = dataset.data, dataset.target

# Train Test Split

In [4]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test =  train_test_split(x, y, test_size = 0.2, random_state = 2024)
x_train, x_val, y_train, y_val =  train_test_split(x_train, y_train, test_size = 0.2, random_state = 2024)

In [ ]:
print(x_train.shape, x_test.shape, x_val.shape)

(113, 13) (36, 13) (29, 13)


In [ ]:
print(y_train)
## 0,1,2의 multiclass label -> one hot encoding 필요

[0 0 2 1 2 2 0 2 0 0 0 1 1 0 1 2 0 0 0 0 0 1 0 2 1 2 1 0 1 0 2 2 1 1 1 1 0
 1 1 1 1 1 1 0 1 1 2 2 0 1 0 1 1 1 0 1 1 0 1 2 1 1 2 0 0 0 1 1 1 1 0 0 2 0
 2 1 1 2 1 0 0 0 1 0 1 2 1 0 0 1 2 1 2 2 0 0 2 2 1 2 0 1 1 1 0 1 1 2 1 0 1
 2 0]


#Train Test set Scailing

In [5]:
from sklearn.preprocessing import StandardScaler

# 1. 스케일러 객체 생성
scaler = StandardScaler()

# 2. Train 데이터로 기준(평균, 표준편차)을 학습(fit)하고 변환(transform)
x_train = scaler.fit_transform(x_train)

# 3. Val과 Test 데이터는 Train의 기준을 그대로 적용하여 변환(transform)만 수행
x_val = scaler.transform(x_val)
x_test = scaler.transform(x_test)

# Multiclass Classification Modeling with PyTorch

## 1.Library Import
- Onehot encoding 함수를 사용하기위해 functional 함수를 함께 import합니다

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import TensorDataset, Dataset, DataLoader
```

In [6]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader

## 2.Dataset / Data Loader 생성/선언

In [7]:
x_train = torch.Tensor(x_train)
x_test = torch.Tensor(x_test)
x_val = torch.Tensor(x_val)

y_train = torch.from_numpy(y_train)
y_test = torch.from_numpy(y_test)
y_val = torch.from_numpy(y_val)

## one hot encoding
y_train = F.one_hot(y_train, num_classes = 3)
y_test = F.one_hot(y_test, num_classes = 3)
y_val = F.one_hot(y_val, num_classes = 3)


batch_size = 16
train_dataset = TensorDataset(x_train, y_train)
val_dataset = TensorDataset(x_val, y_val)
test_dataset = TensorDataset(x_test, y_test)


train_dataloader = DataLoader(train_dataset, batch_size = batch_size, shuffle = True)
val_dataloader = DataLoader(val_dataset, batch_size = batch_size, shuffle = False)
test_dataloader = DataLoader(test_dataset, batch_size = batch_size, shuffle = False)

In [8]:
y_train

tensor([[1, 0, 0],
        [1, 0, 0],
        [0, 0, 1],
        [0, 1, 0],
        [0, 0, 1],
        [0, 0, 1],
        [1, 0, 0],
        [0, 0, 1],
        [1, 0, 0],
        [1, 0, 0],
        [1, 0, 0],
        [0, 1, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 1, 0],
        [0, 0, 1],
        [1, 0, 0],
        [1, 0, 0],
        [1, 0, 0],
        [1, 0, 0],
        [1, 0, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 0, 1],
        [0, 1, 0],
        [0, 0, 1],
        [0, 1, 0],
        [1, 0, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 0, 1],
        [0, 0, 1],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 1, 0],
        [0, 1, 0],
        [0, 0, 1],
        [0, 0, 1],
        [1, 0, 0],
        [0, 1, 0],
        [1, 0, 0],
        [0, 1, 0],
        [0, 

## 3.Model Class 생성 & Model 선언
- Functional Style

In [9]:
class MulticlassModel(nn.Module):
    def __init__(self):
        super(MulticlassModel, self).__init__()

        self.linear_1 = nn.Linear(x_train.shape[1], 32)
        self.relu = nn.ReLU()
        self.linear_2 = nn.Linear(32,64)
        self.linear_3 = nn.Linear(64,128)
        self.linear_4 = nn.Linear(128,y_train.shape[1])
        self.softmax = nn.Softmax(dim = 1)

    def forward(self, x):
        x = self.linear_1(x)
        x = self.relu(x)
        x = self.linear_2(x)
        x = self.relu(x)
        x = self.linear_3(x)
        x = self.relu(x)
        x = self.linear_4(x)

        return x

device = "cuda" if torch.cuda.is_available() else "cpu"
model = MulticlassModel().to(device)

## 4.Optimizer, Loss 선언
- Multiclass Classification Model은 loss function을 CrossEntropyLoss를 사용합니다.

In [10]:
optimizer = optim.Adam(params = model.parameters(), lr = 0.001)

loss_fn = nn.CrossEntropyLoss().to(device)

## 5.Model Training

In [11]:
from sklearn.metrics import accuracy_score

In [12]:
epochs = 100

for epoch in range(epochs):
  train_cost = 0
  train_accuracy = 0
  train_batch_length = len(train_dataloader)

  model.train()
  for train_inputs, train_labels in train_dataloader:

    ## train mode
    train_inputs = train_inputs.to(device)
    train_labels = train_labels.to(device)

    ## Model 예측
    train_preds = model(train_inputs)

    ## loss 계산
    train_loss = loss_fn(train_preds, train_labels.to(float))

    ## optimizer 초기화
    optimizer.zero_grad()

    ## Backpropagation
    train_loss.backward()

    ## weight를 next step으로 이동
    optimizer.step()

    train_cost += train_loss.item()
    train_preds = train_preds.argmax(axis = -1).detach().cpu().numpy()
    train_labels = train_labels.argmax(axis = -1).detach().cpu().numpy()

    train_accuracy += accuracy_score(train_labels, train_preds)

  train_cost /= train_batch_length
  train_accuracy /= train_batch_length

  ## validation set 평가

  val_cost = 0
  val_accuracy = 0
  val_batch_length = len(val_dataloader)

  model.eval() ## layer들을 inference 모드로 동작하도록 설정 변경
  with torch.no_grad(): ## optimizer가 gradient를 계산하는 연산을 수행하지 않고 inference목적으로만 사용하도록 하여 속도를 높임
    for val_inputs, val_labels in val_dataloader:

      val_inputs = val_inputs.to(device)
      val_labels = val_labels.to(device)

      val_preds = model(val_inputs)
      val_loss = loss_fn(val_preds, val_labels.to(float))
      val_cost += val_loss.item()


      val_preds = val_preds.argmax(axis = -1).detach().cpu().numpy()
      val_labels = val_labels.argmax(axis = -1).detach().cpu().numpy()
      val_accuracy += accuracy_score(val_labels, val_preds)

  val_cost /= val_batch_length
  val_accuracy /= val_batch_length

  print("{} / {} epochs => train loss : {:.4f}, train acc : {:.4f}, val loss : {:.4f}, val acc : {:.4f}".format(epoch+1, epochs, train_cost, train_accuracy, val_cost,  val_accuracy))

1 / 100 epochs => train loss : 1.0354, train acc : 0.7578, val loss : 1.0093, val acc : 0.6130
2 / 100 epochs => train loss : 0.9709, train acc : 0.6875, val loss : 0.9188, val acc : 0.6130
3 / 100 epochs => train loss : 0.8238, train acc : 0.8359, val loss : 0.7890, val acc : 0.7139
4 / 100 epochs => train loss : 0.6724, train acc : 0.8750, val loss : 0.6265, val acc : 0.7837
5 / 100 epochs => train loss : 0.4758, train acc : 0.8984, val loss : 0.4512, val acc : 0.8990
6 / 100 epochs => train loss : 0.3205, train acc : 0.9609, val loss : 0.2880, val acc : 0.9688
7 / 100 epochs => train loss : 0.1997, train acc : 0.9844, val loss : 0.1743, val acc : 1.0000
8 / 100 epochs => train loss : 0.1198, train acc : 0.9922, val loss : 0.0910, val acc : 1.0000
9 / 100 epochs => train loss : 0.0717, train acc : 1.0000, val loss : 0.0464, val acc : 1.0000
10 / 100 epochs => train loss : 0.0542, train acc : 1.0000, val loss : 0.0297, val acc : 1.0000
11 / 100 epochs => train loss : 0.0285, train acc

## 6.Model Evaluation

In [13]:
from sklearn.metrics import accuracy_score, classification_report

In [14]:
y_test_pred = model(x_test.to(device))

In [15]:
y_test_pred = y_test_pred.argmax(axis = -1).detach().cpu().numpy()

In [16]:
y_test = y_test.argmax(axis = -1).detach().cpu().numpy()

In [17]:
print(accuracy_score(y_test, y_test_pred))

0.9722222222222222


In [18]:
print(classification_report(y_test, y_test_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       0.93      1.00      0.96        13
           2       1.00      0.92      0.96        12

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



* 모델의 전체 정확도가 기존 36%에서 97%로 상승하였다.
* Class 1으로만 분류하던 편향 문제를 해결하였으며, Class 0은 100%의 Precision과 Recall을 기록하였다.
* 마지막으로, 이렇게 데이터 스케일링(StandardScaler) 적용과 중복된 Softmax 층 제거가 정확하게 작용하여, 모델의 학습 능력을 향상시켰음을 알 수 있다.

